### PR TITLE
Add scala sdk binding

### PR DESCRIPTION
**Pull request by andresgomezfrr**
## _Read then delete this section_

Manual e2e: http://flyteconsole.spotify.net/console/projects/de-gp-pabloc/domains/staging/executions/fbc98a99bb8114ebda7f?duration=all
# TL;DR
_Please replace this text with a description of what this PR accomplishes._

## Type
 - [ ] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
_Remove the '*fixes*' keyword if there will be multiple PRs to fix the linked issue_

fixes https://github.com/flyteorg/flyte/issues/<number>

## Follow-up issue
_NA_
OR
_https://github.com/flyteorg/flyte/issues/<number>_

### CODE REVIEW

File: flytekit-scala-tests/src/test/scala/org/flyte/flytekitscala/SdkScalaTypeTest.scala [hunk 1]
```diff
@@ -17,7 +17,7 @@
 package org.flyte.flytekitscala
 
 import java.time.{Duration, Instant}
-import collection.JavaConverters._
```
Comment (narape) [inline flytekit-scala-tests/src/test/scala/org/flyte/flytekitscala/SdkScalaTypeTest.scala:20]: We are still compiling this with scala 2.12, so let's keep the deprecated library until we drop the support for scala 2.12

File: flytekit-scala_2.13/src/main/scala/org/flyte/flytekit/SdkBindingDataJavaConverters.scala [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.flyte.flytekit
+
+object SdkBindingDataJavaConverters {
```
Comment (narape) [inline flytekit-scala_2.13/src/main/scala/org/flyte/flytekit/SdkBindingDataJavaConverters.scala:19]: Nice

File: flytekit-scala_2.13/src/main/scala/org/flyte/flytekitscala/SdkScalaType.scala [hunk 3]
```diff
@@ -114,93 +114,83 @@ object SdkScalaType {
...
-        }
-
-        value.asInstanceOf[SdkBindingData[T]]
-      }
+      literal =>
+        SdkBindingData.create(
+          toBindingData(literal),
+          sdkLiteral.getLiteralType,
+          sdkLiteral.fromLiteral(literal)
+        )
```
Comment (narape) [inline flytekit-scala_2.13/src/main/scala/org/flyte/flytekitscala/SdkScalaType.scala:163]: Nice here too

File: flytekit-scala_2.13/src/main/scala/org/flyte/flytekitscala/Toolkit.scala [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+  Primitive,
+  Scalar,
+  SimpleType
+}
+import org.flyte.flytekit.{SdkBindingData, SdkWorkflowBuilder}
+
+import java.time.{Duration, Instant}
+import scala.collection.JavaConverters._
+
+object Toolkit {
```
Comment (narape) [inline flytekit-scala_2.13/src/main/scala/org/flyte/flytekitscala/Toolkit.scala:31]: Can you double check if this name `Toolkit` is commonly used in Scala for this sort of things?

File: flytekit-scala_2.13/src/main/scala/org/flyte/flytekitscala/Toolkit.scala [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+    }
+  }
+
+  private def createSdkBindingData[T](value: T): SdkBindingData[T] = {
+    val (bindingData, literalType) = toBindingData(value)
+    SdkBindingData.create(bindingData, literalType, value)
+  }
+
+  implicit def stringSdkBinding(string: String): SdkBindingData[String] =
+    createSdkBindingData(string)
```
Comment (narape) [inline flytekit-scala_2.13/src/main/scala/org/flyte/flytekitscala/Toolkit.scala:123]: Magic

**Review by narape — COMMENTED** [review]
Only one real comment about keep using a deprecated library, one about to check with other libraries in Scala if Toolkit is commonly used and some praise comments

File: flytekit-scala_2.13/src/main/scala/org/flyte/flytekitscala/Toolkit.scala (hunk 4 above)
Comment (andresgomezfrr) [reply flytekit-scala_2.13/src/main/scala/org/flyte/flytekitscala/Toolkit.scala:25]: mmmm yeah! you are right... I think that maybe `Implicits` is better... I will change it.

File: flytekit-scala-tests/src/test/scala/org/flyte/flytekitscala/SdkScalaTypeTest.scala (hunk 1 above)
Comment (andresgomezfrr) [reply flytekit-scala-tests/src/test/scala/org/flyte/flytekitscala/SdkScalaTypeTest.scala:20]: 100% agree!

**Review by narape — APPROVED** [review]

File: flytekit-scala_2.13/src/main/scala/org/flyte/flytekitscala/SdkBindingData.scala [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+  def ofDateTime(instant: Instant): SdkJavaBindinigData[Instant] =
+    createSdkBindingData(instant)
+
+  def ofDuration(
+      duration: Duration
+  ): SdkJavaBindinigData[Duration] = createSdkBindingData(duration)
+
+  def ofCollection[T](
+      collection: List[T]
+  ): SdkJavaBindinigData[List[T]] = createSdkBindingData(collection)
```
Comment (narape) [inline flytekit-scala_2.13/src/main/scala/org/flyte/flytekitscala/SdkBindingData.scala:55]: We have the problem of empty list that we had in Java and deferring users to `SdkBindingData.create(...)` is a too heavy toll for that. 
In java we added methods to pass a LiteralType or add more factory methods that know their literaltype

**Review by narape — COMMENTED** [review]
Please also add a description to the PR

File: flytekit-scala_2.13/src/main/scala/org/flyte/flytekitscala/SdkBindingData.scala (hunk 6 above)
Comment (andresgomezfrr) [reply flytekit-scala_2.13/src/main/scala/org/flyte/flytekitscala/SdkBindingData.scala:55]: I'm going to check it!

Comment (andresgomezfrr) [reply flytekit-scala_2.13/src/main/scala/org/flyte/flytekitscala/SdkBindingData.scala:55]: Done! 👍🏼

File: flytekit-scala_2.13/src/main/scala/org/flyte/flytekitscala/SdkBindingData.scala [hunk 7]
```diff
@@ -54,11 +54,160 @@ object SdkBindingData {
       collection: List[T]
   ): SdkJavaBindinigData[List[T]] = createSdkBindingData(collection)
 
+  def ofCollection[T](
+      collection: List[T],
+      literalType: LiteralType
+  ): SdkJavaBindinigData[List[T]] =
```
Comment (narape) [inline flytekit-scala_2.13/src/main/scala/org/flyte/flytekitscala/SdkBindingData.scala:60]: We have this flipped in Java, type first and then list. We should be consistent

**Review by narape — COMMENTED** [review]
A small comment about consistency in Java and Scala

**Pull request merged**